llama-cpp-2/src/mtmd.rs

```diff
@@ -0,0 +1,899 @@
+        }
+    }
+}
+
+impl From<&MtmdContextParams> for llama_cpp_sys_2::mtmd_context_params {
+    fn from(params: &MtmdContextParams) -> Self {
+        let mut context = unsafe { llama_cpp_sys_2::mtmd_context_params_default() };
+
+        context.use_gpu = params.use_gpu;
+        context.print_timings = params.print_timings;
+        context.n_threads = params.n_threads;
+        context.media_marker = params.media_marker.as_ptr();
+
+        context
+    }
+}
+
+/// Text input configuration
+///
+/// # Examples
+///
+/// ```
+/// use llama_cpp_2::mtmd::MtmdInputText;
+///
+/// let input = MtmdInputText {
+///     text: "Describe this image.".to_string(),
+///     add_special: true,
+///     parse_special: true,
+/// };
+/// ```
+#[derive(Debug, Clone)]
+pub struct MtmdInputText {
+    /// The input text string
+    pub text: String,
+    /// Whether to add special tokens
+    pub add_special: bool,
+    /// Whether to parse special tokens
+    pub parse_special: bool,
+}
+
+/// Safe wrapper around `mtmd_context`.
+///
+/// This represents an initialized multimodal context that can process
+/// text, images, and audio through llama.cpp's multimodal interface.
+#[derive(Debug)]
+pub struct MtmdContext {
+    pub(crate) context: NonNull<llama_cpp_sys_2::mtmd_context>,
+}
+
+impl MtmdContext {
+    /// Initialize MTMD context from a multimodal projection file.
+    ///
+    /// # Arguments
+    ///
+    /// * `mmproj_path` - Path to the multimodal projection file
+    /// * `text_model` - Reference to the text model
+    /// * `params` - Configuration parameters for the MTMD context
+    ///
+    /// # Returns
+    ///
+    /// Returns `Ok(MtmdContext)` on success, or `Err(MtmdInitError)` on failure.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if:
+    /// - The path cannot be converted to a C string
+    /// - The underlying C function returns null (indicating initialization failure)
+    pub fn init_from_file(
+        mmproj_path: &str,
+        text_model: &LlamaModel,
+        params: &MtmdContextParams,
+    ) -> Result<Self, MtmdInitError> {
+        let path_cstr = CString::new(mmproj_path)?;
+        let ctx_params = llama_cpp_sys_2::mtmd_context_params::from(params);
+
+        let context = unsafe {
+            llama_cpp_sys_2::mtmd_init_from_file(
+                path_cstr.as_ptr(),
+                text_model.model.as_ptr(),
+                ctx_params,
+            )
+        };
+
+        if context.is_null() {
+            return Err(MtmdInitError::NullResult);
+        }
+
+        let context = NonNull::new(context).ok_or(MtmdInitError::NullResult)?;
+        Ok(Self { context })
+    }
+
+    /// Check whether non-causal attention mask is needed before `llama_decode`.
+    #[must_use] pub fn decode_use_non_causal(&self) -> bool {
+        unsafe { llama_cpp_sys_2::mtmd_decode_use_non_causal(self.context.as_ptr()) }
+    }
```

Comment:
a safety comment would be great. (self.context is not null should be sufficent?)